pkg/controllers/securitypolicy_controller.go

```diff
@@ -44,45 +46,56 @@ type SecurityPolicyReconciler struct {
 func (r *SecurityPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	obj := &v1alpha1.SecurityPolicy{}
 	log.Info("reconciling securitypolicy CR", "securitypolicy", req.NamespacedName)
+	metrics.ControllerSyncTotal.WithLabelValues(RES_TYPE).Inc()
 
 	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
 		log.Error(err, "unable to fetch security policy CR", "req", req.NamespacedName)
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
 	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		metrics.ControllerUpdateTotal.WithLabelValues(RES_TYPE).Inc()
 		if !controllerutil.ContainsFinalizer(obj, util.FinalizerName) {
 			controllerutil.AddFinalizer(obj, util.FinalizerName)
 			if err := r.Client.Update(ctx, obj); err != nil {
+				metrics.ControllerUpdateFailTotal.WithLabelValues(RES_TYPE).Inc()
 				return ctrl.Result{}, err
 			}
 			log.V(1).Info("added finalizer on securitypolicy CR", "securitypolicy", req.NamespacedName)
 		}
 
 		if isCRInSysNs, err := r.isCRRequestedInSystemNamespace(&ctx, &req); err != nil {
+			metrics.ControllerUpdateFailTotal.WithLabelValues(RES_TYPE).Inc()
 			return ctrl.Result{}, err
 		} else if isCRInSysNs {
 			err = errors.New("Security Policy CR cannot be created in System Namespace")
 			log.Error(err, "failed to create security policy CR", "securitypolicy", req.NamespacedName)
+			metrics.ControllerUpdateFailTotal.WithLabelValues(RES_TYPE).Inc()
 			return ctrl.Result{}, err
 		}
 
 		if err := r.Service.CreateOrUpdateSecurityPolicy(obj); err != nil {
 			log.Error(err, "failed to create or update security policy CR", "securitypolicy", req.NamespacedName)
 			r.setSecurityPolicyReadyStatusFalse(&ctx, obj, &err)
+			metrics.ControllerUpdateFailTotal.WithLabelValues(RES_TYPE).Inc()
 			return ctrl.Result{}, err
 		}
 		r.setSecurityPolicyReadyStatusTrue(&ctx, obj)
+		metrics.ControllerUpdateSuccessTotal.WithLabelValues(RES_TYPE).Inc()
 	} else {
 		if containsString(obj.GetFinalizers(), util.FinalizerName) {
+			metrics.ControllerDeleteTotal.WithLabelValues(RES_TYPE).Inc()
 			if err := r.Service.DeleteSecurityPolicy(obj.UID); err != nil {
+				metrics.ControllerDeleteFailTotal.WithLabelValues(RES_TYPE).Inc()
 				return ctrl.Result{}, err
 			}
 			controllerutil.RemoveFinalizer(obj, util.FinalizerName)
 			if err := r.Client.Update(ctx, obj); err != nil {
+				metrics.ControllerDeleteFailTotal.WithLabelValues(RES_TYPE).Inc()
 				return ctrl.Result{}, err
```

Comment:
I think in metrics.go, we can define some functions to update the values, for example, update_counter. In the function, we can also add check for if vmc env.

Reply:
got it. I'll try it.